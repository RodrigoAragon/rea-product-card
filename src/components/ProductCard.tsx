import React, { JSX }  from "react"
import { createContext, CSSProperties } from "react";
import { InitialValues, onChangeArgs, Product, ProductCardHandlers, ProductContextProps } from '../interfaces/Products.interfaces';
import { useProduct } from "../hooks/useProduct";

import styles from "../styles/styles.module.css"

export const ProductContext = createContext({} as ProductContextProps)
const {Provider} = ProductContext


export interface Props{
  children?: (args: ProductCardHandlers) => JSX.Element
  className?: string
  product: Product
  style?: CSSProperties
  onChange?: (args:onChangeArgs) => void
  value?: number
  initialValues?: InitialValues
}


export const ProductCard = ({children, product, className, style, onChange, value, initialValues}:Props) => {

  const {counter, increase, maxCount, isMaxCountReached, reset} =  useProduct({onChange, product, value, initialValues});

  return (
    <Provider 
      value={{
        counter,
        increase,
        maxCount,
        product
      }}
    >
      <div className={`${styles.productCard} ${className}`} style={style}>

      {
        children && children({
          count: counter,
          isMaxCountReached,
          maxCount: initialValues?.maxCount,
          product,

          increase,
          reset,
      })
      }

      </div>
    </Provider>
    
  )
}