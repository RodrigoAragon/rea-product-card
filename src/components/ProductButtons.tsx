import React  from "react"
import { CSSProperties, useCallback, useContext } from "react"
import { ProductContext } from "./ProductCard"
import styles from "../styles/styles.module.css"

export interface Props{
  className?:string
  style?: CSSProperties
}

export const ProductButtons = ({className, style}:Props) =>{
    const {counter, increase, maxCount} = useContext(ProductContext)
    const isMaxReached = useCallback(
      () => {
        return !!maxCount && counter === maxCount? true : false 
      },
      [counter, maxCount],
    )
  
  return(
    <div className={`${styles.buttonsContainer} ${className}`} style={style}>
      <button 
        className={styles.buttonMinus} 
        onClick={() => increase(-1)}> - </button>
  
      <div className={styles.countLabel}> {counter} </div>
  
      <button 
        className={isMaxReached()? `${styles.disabled}` :  `${styles.buttonAdd}`} 
        onClick={() => increase(+1)}> + </button>
    </div>
  )
}