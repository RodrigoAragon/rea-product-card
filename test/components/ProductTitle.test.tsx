import React from "react";
import {render} from "@testing-library/react"
import { ProductTitle } from '../../src/components/ProductTitle';
import { ProductCard } from "../../src/components/ProductCard";
import { product1 } from '../data/products';

describe('Pruebas en el Product Title', () => {

    it('Debe mostrar el componente correctamente con el titulo personalizado', () => {
        const wrapper = render(<ProductTitle title="Custom Product"/>)

        expect(wrapper).toMatchSnapshot()
    });

    it('Debe mostrar el componente correctamente con el nombre del producto', () => {
        const wrapper = render(
        <ProductCard product={product1}>
            {
                () => (
                    <ProductTitle/>
                )
            }
        </ProductCard>
    )

        expect(wrapper).toMatchSnapshot()
    });
})