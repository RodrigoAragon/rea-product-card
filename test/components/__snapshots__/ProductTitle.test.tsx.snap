// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el Product Title Debe mostrar el componente correctamente con el nombre del producto 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="productCard undefined"
      >
        <span
          class="productDescription undefined"
        >
          Producto número 1
        </span>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="productCard undefined"
    >
      <span
        class="productDescription undefined"
      >
        Producto número 1
      </span>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Pruebas en el Product Title Debe mostrar el componente correctamente con el titulo personalizado 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <span
        class="productDescription undefined"
      >
        Custom Product
      </span>
    </div>
  </body>,
  "container": <div>
    <span
      class="productDescription undefined"
    >
      Custom Product
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
